[llvm][CodeGen] Added a check in CodeGenPrepare::optimizeSwitchType

karouzakisp opened the pull request:
When we have two or more users in the Switch Condition and we have more SExt Users than ZExt Users or more signed cmp Instruction Users than unsigned cmp Instruction Users, we need to check to not insert a ZExt into the switch condition Type and rather insert a SExt. This way we can eliminate one Instruction.

arsenm commented:
Can you rephrase the description to say what the check is, and why?

preames commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -7397,7 +7397,19 @@ bool CodeGenPrepare::optimizeSwitchType(SwitchInst *SI) {
       ExtType = Instruction::ZExt;
   }
-
+  // If Cond has 2 users and one of them is a sext don't use a zext.
```
We should probably be looking at the other user and using it's extend kind (sign or zero).

we probably do want the argument extend kind check to overrule, so maybe put this just above?

preames commented on llvm/test/CodeGen/RISCV/switch-width.ll:
```diff
@@ -194,25 +194,108 @@ return:
   %retval = phi i32 [ -1, %sw.default ], [ 0, %sw.bb0 ], [ 1, %sw.bb1 ]
   ret i32 %retval
 }
+define i32 @sext_i32(i16 %a)  {
```
You should add test variants here with zeroext and signext on the arguments to make sure we test the interaction.

karouzakisp committed:
[llvm][CodeGen] Added a check in CodeGenPrepare::OptimizeSwitchType,
so that if we have more SExt Users than ZExt Users we use SExt
instead of a ZExt in the Switch Condition
because that creates a redundant instruction.

arsenm commented on llvm/lib/CodeGen/CodeGenPrepare.cpp:
```diff
@@ -7387,17 +7387,38 @@ bool CodeGenPrepare::optimizeSwitchType(SwitchInst *SI) {
   }
 
+  // Some targets prefer SExt over ZExt.
+  if (TLI->isSExtCheaperThanZExt(OldVT, RegType))
```
What was the point of doing this user scan if you just throw away the result?

karouzakisp replied:
Because in some cases we have more ZExtUsers but SExt is cheaper so if create a ZExt we add more instructions. 
I checked spec2017 for RISC-V with the precedence with this and It was doing worse on 500.perlbench_r if we check SExtIsCheaperThanZExt before the user scan. I can share the results later

karouzakisp replied:
This is the text sizes with SExtIsCheaperThanZExt take no precedence over the user scan 
![SextIsCheaperThanZextisBeforeUserScan](https://github.com/llvm/llvm-project/assets/45971450/3608815f-28d3-483d-be3a-21d61b0340c9)
And this is the text sizes with SExtIsCheaperThanZExt taking precedence than user scan.
![SextIsCheaperThanZextisAfterUserScan](https://github.com/llvm/llvm-project/assets/45971450/d4809dba-67f9-4087-84da-0c4121030d62)

arsenm replied:
I don't mean the end result, I mean the computation you perform in this code is discarded based on isSExtCheaperThanZExt

karouzakisp replied:
Yes isSExtCheaperThanZExt needs to take precedence if it doesn't agree with the computation that I perform otherwise the code size increases as I showed above.
Maybe the gain is very small compared to the cost because we need a User Scan.

michaelmaitland replied:
My reading of @arsenm's comment is that maybe it would be better to avoid the user scan if we dont need it as such:
```
if (TLI->isSExtCheaperThanZExt(OldVT, RegType))
  ExtType = Instruction::SExt;
else {
 // User scan above
}

topperc replied:
If one of the users is a ZExt, then ZExt is preferable to SExt even if SExt is cheaper.

arsenm reviewed (commented):
Title should be more specific about the purpose of the check, rather than state that a check was added

arsenm reviewed (changes requested):
Title needs improvement
